src/java.base/share/classes/java/lang/StableValue.java

```diff
@@ -0,0 +1,664 @@
+ *    // @link substring="of" target="#of" :
+ *    private final StableValue<Logger> logger = StableValue.of();
+ *
+ *    private Logger getLogger() {
+ *        return logger.orElseSet( () -> Logger.create(Component.class) );
+ *    }
+ *
+ *    public void process() {
+ *        getLogger().info("Process started");
+ *        // ...
+ *    }
+ * }
+ *}
+ * <p>
+ * The {@code getLogger()} method calls {@code logger.orElseSet()} on the stable value to
+ * retrieve its content. If the stable value is <em>unset</em>, then {@code orElseSet()}
+ * evaluates and sets the content; the content is then returned to the client. In other
+ * words, {@code orElseSet()} guarantees that a stable value's content is <em>set</em>
+ * before it is used.
+ * <p>
+ * Furthermore, {@code orElseSet()} guarantees that the supplier provided is
+ * evaluated only once, even when {@code logger.orElseSet()} is invoked concurrently.
+ * This property is crucial as evaluation of the supplier may have side effects,
+ * e.g., the call above to {@code Logger.create()} may result in storage resources
+ * being prepared.
+ *
+ * <h2 id="stable-functions">Stable Functions</h2>
+ * Stable values provide the foundation for higher-level functional abstractions. A
+ * <em>stable supplier</em> is a supplier that computes a value and then caches it into
+ * a backing stable value storage for later use. A stable supplier is created via the
+ * {@linkplain StableValue#supplier(Supplier) StableValue.supplier()} factory, by
+ * providing an original {@linkplain Supplier} which is invoked when the stable supplier
+ * is first accessed:
+ *
+ * {@snippet lang = java:
+ * public class Component {
+ *
+ *     private final Supplier<Logger> logger =
+ *             // @link substring="supplier" target="#supplier(Supplier)" :
+ *             StableValue.supplier( () -> Logger.getLogger(Component.class) );
+ *
+ *     public void process() {
+ *        logger.get().info("Process started");
+ *        // ...
+ *     }
+ * }
+ *}
+ * A stable supplier encapsulates access to its backing stable value storage. This means
+ * that code inside {@code Component} can obtain the logger object directly from the
+ * stable supplier, without having to go through an accessor method like {@code getLogger()}.
+ * <p>
+ * A <em>stable int function</em> is a function that takes an {@code int} parameter and
+ * uses it to compute a result that is then cached by the backing stable value storage
+ * for that parameter value. A stable int function is created via the
+ * {@linkplain StableValue#intFunction(int, IntFunction) StableValue.intFunction()}
+ * factory. Upon creation, the input range (i.e. [0, size)) is specified together with
+ * an original {@linkplain IntFunction} which is invoked at most once per input value. In
+ * effect, the stable int function will act like a cache for the original {@linkplain IntFunction}:
+ *
+ * {@snippet lang = java:
+ * public final class SqrtUtil {
```

Comment:
Maybe this example can be improved? E.g. how about a general utility class which speeds up computation of common square roots, up to a certain number. Then you can have a method that takes a number and returns its square root (it will check if the number is less than the threshold and use the stable function cache in that case). E.g. this would be very nice with primitive type patterns:

```
if (input instanceof int i && i < threshold) {
   return SQRT.apply(i);
} else {
   return Math.sqrt(input);
}
```

(it's probably not great to refer to another preview here -- maybe we can keep this example for another time). But, if you use something like `pow`, that might be easier (as then you can only worry about int/long inputs) ?